### ISSUE TITLE
Returning Ref<T> in virtual methods

### ISSUE THREAD

**Issue by BastiaanOlij** [godotengine/godot-cpp#954]
After doing a bunch of research into this in the last couple days I figured I detail out a bit of the conclusions around this in an issue here for prosperity. Right now this info is spread out over the Godot Contributors Chat GDExtension channel.

First off, #801 gives some great background information and a valid workaround fix, apologies to @zhehangd that he did all this work for us to only dive into it properly now.

It is important to realize that the object class `Ref<T>` is not published through GDExtensions, the implementation of `Ref<T>` on the godot-cpp side is a separate (but nearly identical) implementation of this class. Only object pointers of class T are exchanged and it is in the `Ref<T>` implementation on the godot-cpp side that reference counting is handled.

When we look at "normal" script calls into GDExtension, all parameters, including the return parameter, all are encoded into Variants and thus only variants are exchanged between Godot and the extension. When an object subclassed from `RefCounted` is passed through a variant, the variant class handles reference counting and thus also properly increases/decreases the refcount on the object.

# Godot virtual calls

When we look at virtual method calls these are different as here we wanted to support any datatype including custom structs and we thus can't rely on variants.
This code is a little bit more hard to follow because we're looking at generated macros here, so bear with me. We pick up the thread on the Godot side in `core/object/gdvirtual.gen.inc` (generated so it's not in the source tree until you compile). The macro `GDVIRTUAL1R` is one of a number of macros that defines our virtual call and most importantly generates the method `bool _gdvirtual_##m_name##_call(m_type1 arg1, m_ret& r_ret)`.

This code is used both for calling methods implemented in GDScript, which does use Variants (as GDScript can only handle variants, I'm not entirely sure how this will react if unsupported datatypes are used but that's not in scope of this issue).

The important code comes next as we handle calling a virtual method implemented in a GDExternal (cleaned up a bit for readability):
```
PtrToArg<m_type1>::EncodeT argval1 = arg1;
GDNativeConstTypePtr argptrs[1]={&argval1};

PtrToArg<m_ret>::EncodeT ret;
_gdvirtual_##m_name(_get_extension_instance(),reinterpret_cast<GDNativeConstTypePtr*>(argptrs),&ret);
r_ret = (m_ret)ret;
return true;
```

The important bit here is `PtrToArg<T>::EncodeT` which provides a mechanism for storing the data we are exchanging and some compiler magic to convert between the original type and the storage type. The actual call to the method inside of the GDExtension external then takes pointers to this storage (i.e. `argptrs` which is an array of pointers to our arguments and `&ret`).

So let's take a look at our relevant `PtrToArg` class (https://github.com/godotengine/godot/blob/master/core/object/ref_counted.h#L253):
```
template <class T>
struct PtrToArg<Ref<T>> {
	_FORCE_INLINE_ static Ref<T> convert(const void *p_ptr) {
		return Ref<T>(const_cast<T *>(reinterpret_cast<const T *>(p_ptr)));
	}

	typedef Ref<T> EncodeT;

	_FORCE_INLINE_ static void encode(Ref<T> p_val, const void *p_ptr) {
		*(const_cast<Ref<RefCounted> *>(reinterpret_cast<const Ref<RefCounted> *>(p_ptr))) = p_val;
	}
};
```

So first thing that stands out here: `typedef Ref<T> EncodeT;`, so we are using `Ref<T>` as our storage and thus just duplicating the reference. We're thus sending pointers to our `Ref<T>` variables to GDExternal even though GDExternal _has no idea_ about Godots `Ref<T>` class.

Our two methods `encode` and `convert` are no used here, they are used in other cases. What stands out here is that `encode` takes a `Ref<T>` and "encodes" it into a `const void *`, this `const void *` is assumed to be a pointer to a `Ref<T>` variable.
HOWEVER `convert` is assuming our `const void *` is a pointer to `T`!

There is no further documentation here on why this is so. 

What is important here is that our call to `_gdvirtual_##m_name` ends up sending pointers to `Ref<T>` variables both for parameters and our return value.

# Godot-cpp virtual calls

So let's have a look at what is happening on the Godot-cpp side...

Here our entry point is created with the macro `BIND_VIRTUAL_METHOD`. Looking at this macro we can simply see that this creates a function with the appropriate name and registers this method with ClassDB:
```
#define BIND_VIRTUAL_METHOD(m_class, m_method)
	{
		auto ___call##m_method = [](GDNativeObjectPtr p_instance, GDNativeConstTypePtr *p_args, GDNativeTypePtr p_ret) -> void {
			call_with_ptr_args(reinterpret_cast<m_class *>(p_instance), &m_class::m_method, p_args, p_ret);
		};
		godot::ClassDB::bind_virtual_method(m_class::get_class_static(), #m_method, ___call##m_method);
	}
```

The important bit here is calling `call_with_ptr_args`, these are 4 template methods defined in `binder_common.hpp`, the 4 variations are based on whether the method we're calling is defined as const and whether we have a return value.

The real work happens in `call_with_ptr_args_ret_helper`, specifically:
```
PtrToArg<R>::encode((p_instance->*p_method)(PtrToArg<P>::convert(p_args[Is])...), r_ret);
```

This calls our actual method and what is important here is that we're calling `PtrToArg<P>::convert` for all our parameters, and we're calling PtrToArg<R>::encode` to copy our methods return value into our return variable (which in our case is a pointer to a **GODOT** `Ref<T>` variable).

Now let's assume the method we are calling is defined as: `Ref<ClassA> test_method(Ref<ClassB> p_parameter)`. So both input and output are **godot-cpp** `Ref<T>` objects. Let's have a look at our `PtrToArg` class (we find this in ref.hpp):

```
template <class T>
struct PtrToArg<Ref<T>> {
	_FORCE_INLINE_ static Ref<T> convert(const void *p_ptr) {
		return Ref<T>(reinterpret_cast<T *>(godot::internal::gdn_interface->object_get_instance_binding(*reinterpret_cast<GDNativeObjectPtr *>(const_cast<void *>(p_ptr)), godot::internal::token, &T::___binding_callbacks)));
	}

	typedef Ref<T> EncodeT;

	_FORCE_INLINE_ static void encode(Ref<T> p_val, void *p_ptr) {
		*reinterpret_cast<const GodotObject **>(p_ptr) = p_val->_owner;
	}
};
```

Looks relatively similar right? And indeed it looks copied but we find some weirdness here.

> note a quick note here, asfar as I can see `PtrToArg<Ref<T>>::convert` and `PtrToArg<Ref<T>>::encode` are only used when calling our virtual functions and when calling static functions (in `call_with_ptr_args_static_method_helper` and `call_with_ptr_args_static_method_ret_helper`).

First `convert`, just like its **Godot** counterpart it is casting `T*`, and it is thus assuming 'p_ptr' is pointing to a `RefCounted`, not to `Ref<T>` variable.
**We don't have examples of this yet but I think this is thus broken**.

The one related to the reported issues however is `encode` and here something very silly happens. At first glance `GodotObject` stands out and you'd be remiss to think we have the same issue as in `convert`, however notice the double `*`.

To understand what is happening here we need to look at `Ref<T>` on the **Godot** side:
```
template <class T>
class Ref {
	T *reference = nullptr;

	...
```
Note the first member of `Ref<T>` is our `reference` variable which is a pointer to `T`.

As `Ref` has no virtual members and we thus do not have a symbols table, our `p_ptr` thus also points to `p_ptr->reference` and thus casting to `**` allows us to modify the contents of `p_ptr->reference`.
We're thus completely bypassing the setter and thus bypassing our logic that will increase our refcount, hence the early destruction.

The workaround @zhehangd suggested would thus fix this but arguably the whole approach is wrong as any change in how `Ref<T>` is implemented in Godot would break the solution.

# Solutions

There are a few options to consider here.

## Accept the workaround

It is totally valid to accept zhehangd workaround though I would add a lot more commentary to indicate why we're doing what we're doing and the inherit danger within.

We should also fix `convert` and add clear comments as to why we're expecting a **Godot** `Ref<T>` variable.

## Add a T getter and setter for Godot Ref.

A slightly nicer version then the above is to add a `typedef void *GDNativeRefPtr` and add `typedef GDNativeObjectPtr (*GDNativeRefGetter)(GDNativeRefPtr p_ref)` and `typedef void (*GDNativeRefSetter)(GDNativeRefPtr p_ref, GDNativeObjectPtr p_obj)` as interface methods to get/set the object on a **Godot** `Ref<T>` object.

## Change our GDVIRTUAL... implementation

We can look into changing `GDVIRTUAL...` to not use pointers to `Ref<T>` but just use `<T>`. We're going to run into trouble with this as 'ArgToPtr' is used in other places here it is already assumed that parameters are passed as pointers to `T` but the return value is a pointer to `Ref<T>`.

Alternatively we could look into using Variants here somehow.

**Comment by groud** [comment]
We discussed this issue during the PR review meeting, and it seems like solution 2 is likely the most adequate. It does not change anything for godot-cpp users and only modify `Ref<T>`'s API for internal use. So that should be fine.

**Comment by zhehangd** [comment]
Adding special interface methods for RefCounted objects sounds good to me, as it creates a clear context to transfer the ownership between different Ref wrappers. Other language bindings may benefit from it as well.

However be careful that one can have a function returning a `Ref<RefCounted>` and also a function returning a `RefCounted*`. For the first case, we certainly want the engine to manage the object using engine's `Ref`. For the second case, I don't exactly remember how godot/godot-cpp deal with it, but it seems that returning a fresh memnew raw pointer does work and we can use it in GDScript as if it is properly managed. To complicate it more, what if one returns a raw RefCounted pointer which have been referenced several times (for example by some internal Ref objects)? In any of these cases, should we use `GDNativeRefSetter` or not? I guess we need to make sure the behavior is clear enough (since returning a raw RefCounted ptr is so confusing I don't mind if it triggers a compiling error / static assertion).

**Comment by BastiaanOlij** [comment]
@zhehangd this is indeed allowed but what happens here really depends on the calling method.
Generally speaking there shouldn't be a situation where it makes sense that a method is called where we expect an `Object *` to be returned, that the implementation of that method returns a `RefCounted *` as we can't expect the calling method to handle refcounting. However since RefCounted is a subclass from Object its theoretically possible.

What is possible and valid situation is where the method is designed to return a `RefCounted *` but in this case the expectation is that the calling method handles this. Note also that the normal GDExtension calls always use Variants as intermediates and Variants always apply refcounting when a subclass of RefCounted is passed closing that loophole.
Virtual calls however do not, there is a bigger expectation of developers doing the right thing.

Personally I see it purely as a technical possibility, not good programming. IMHO any object instantiated from RefCounted and it's subclasses should always be handled through Ref<T>, so any virtual method that expects a RefCounted object to be passed to it or returned, should be using Ref<T>. Arguments could be pointers as it's guaranteed the object will live for the duration of the call, but return values should never be RefCounted pointers.

**Comment by dsnopek** [comment]
This was fixed by https://github.com/godotengine/godot-cpp/pull/958
